src/hardware/pcspeaker_impulse.cpp

```diff
@@ -0,0 +1,543 @@
+{
+#ifdef SPKR_DEBUGGING
+	LOG_INFO("PCSPEAKER: %f output: %s, clock gate %s",
+	         PIC_FullIndex(),
+	         port_b_state.speakerOutput ? "pit" : "forced low",
+	         port_b_state.timerGateOutput ? "on" : "off");
+#endif
+	const auto new_index = static_cast<float>(PIC_TickIndex());
+	ForwardPIT(new_index);
+	// pit clock gate enable rising edge is a trigger
+	const bool pit_trigger = !prev_port_b_state.timerGateOutput && port_b_state.timerGateOutput;
+
+	prev_port_b_state.data = port_b_state.data;
+	if (pit_trigger) {
+		switch (pit.mode) {
+		case PitMode::OneShot:
+			if (pit.mode1_waiting_for_counter) {
+				// assert output amplitude is high
+				break;
+			}
+			pit.amplitude = negative_amplitude;
+			pit.index     = 0;
+			pit.max_ms    = pit.mode1_pending_max;
+
+			pit.mode1_waiting_for_trigger = 0;
+			break;
+
+		case PitMode::SquareWave:
+		case PitMode::SquareWaveAlias:
+			pit.mode3_counting = true;
+			// pit.new_max_ms = pit.new_max_ms; // typo or bug?
+			pit.index       = 0;
+			pit.max_ms      = pit.new_max_ms;
+			pit.new_half_ms = pit.new_max_ms / 2;
+			pit.half_ms     = pit.new_half_ms;
+			pit.amplitude   = positive_amplitude;
+			break;
+
+		default:
+			// TODO: implement other modes
+			break;
+		}
+	} else if (!port_b_state.timerGateOutput) {
+		switch (pit.mode) {
+		case PitMode::OneShot:
+			// gate amplitude does not affect mode1
+			break;
+
+		case PitMode::SquareWave:
+		case PitMode::SquareWaveAlias:
+			// low gate forces pit output high
+			pit.amplitude      = positive_amplitude;
+			pit.mode3_counting = false;
+			break;
+
+		default:
+			// TODO: implement other modes
+			break;
+		}
+	}
+	if (port_b_state.speakerOutput) {
+		AddImpulse(new_index, pit.amplitude);
+	} else {
+		AddImpulse(new_index, negative_amplitude);
+	}
+}
+
+// TODO: check if this is accurate
+static double sinc(const double t)
+{
+	double result = 1.0;
+
+	constexpr auto SINC_ACCURACY = 20;
+	for (auto k = 1; k < SINC_ACCURACY; ++k) {
+		result *= cos(t / pow(2.0, k));
+	}
+	return result;
+}
+
+float PcSpeakerImpulse::CalcImpulse(const double t)
+{
+	// raised-cosine-windowed sinc function
+	const double fs = sample_rate;
+	const auto fc   = fs / (2 + static_cast<double>(cutoff_margin));
+	const auto q    = static_cast<double>(filter_quality);
+	if ((0 < t) && (t * fs < q)) {
+		const auto window    = 1.0 + cos(2 * fs * M_PI * (q / (2 * fs) - t) / q);
+		const auto amplitude = window * (sinc(2 * fc * M_PI * (t - q / (2 * fs)))) / 2.0;
+		return static_cast<float>(amplitude);
+	} else
+		return 0.0f;
+}
+
+void PcSpeakerImpulse::AddImpulse(float index, int16_t amplitude)
+{
+	// Did the amplitude change?
+	if (amplitude == pit.prev_amplitude)
+		return;
+
+	pit.prev_amplitude = amplitude;
+
+	// Wake the channel if it was sleeping
+	if (!channel->is_enabled)
+		channel->Enable(true);
+
+	// Make sure the time index is valid
+	index = clamp(index, 0.0f, 1.0f);
+#ifndef REFERENCE
```

Comment:
What's this?

Reply:
Added clarification that this is the fast implementation that uses per-computed sinc lookup tables (versus the very slow reference implementation).